channels/rdpdr/client/rdpdr_capabilities.c

```diff
@@ -120,53 +157,77 @@ static void rdpdr_write_smartcard_capset(rdpdrPlugin* rdpdr, wStream* s)
 }
 
 /* Process smartcard redirection capability set */
-static void rdpdr_process_smartcard_capset(rdpdrPlugin* rdpdr, wStream* s)
+static UINT rdpdr_process_smartcard_capset(rdpdrPlugin* rdpdr, wStream* s)
 {
 	UINT16 capabilityLength;
 
+	if (Stream_GetRemainingLength(s) < 2)
+		return CHANNEL_RC_NO_BUFFER;
+
 	Stream_Read_UINT16(s, capabilityLength);
+
+	if (Stream_GetRemainingLength(s) < capabilityLength - 4)
+		return CHANNEL_RC_NO_BUFFER;
+
 	Stream_Seek(s, capabilityLength - 4);
+
+	return CHANNEL_RC_OK;
 }
 
-void rdpdr_process_capability_request(rdpdrPlugin* rdpdr, wStream* s)
+UINT rdpdr_process_capability_request(rdpdrPlugin* rdpdr, wStream* s)
 {
+	UINT status = CHANNEL_RC_OK;
 	UINT16 i;
 	UINT16 numCapabilities;
 	UINT16 capabilityType;
 
+	if (!rdpdr || !s)
+		return CHANNEL_RC_NULL_DATA;
+
+	if (Stream_GetRemainingLength(s) < 4)
+		return CHANNEL_RC_NO_BUFFER;
+
 	Stream_Read_UINT16(s, numCapabilities);
 	Stream_Seek(s, 2); /* pad (2 bytes) */
 
+	if (Stream_GetRemainingLength(s) < sizeof(UINT16) * numCapabilities)
+		return CHANNEL_RC_NO_BUFFER;
```

Comment:
there's possibly an int overflow here, should be:
if (Stream_GetRemainingLength(s) / sizeof(UINT16) < numCapabilities )